sql/src/main/java/org/apache/druid/sql/calcite/external/TableAppendMacro.java

```diff
@@ -0,0 +1,344 @@
+import org.apache.calcite.sql.type.SqlOperandMetadata;
+import org.apache.calcite.sql.type.SqlTypeFamily;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.calcite.sql.validate.SqlUserDefinedTableMacro;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql.validate.SqlValidatorCatalogReader;
+import org.apache.calcite.sql.validate.SqlValidatorTable;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.query.DataSource;
+import org.apache.druid.query.UnionDataSource;
+import org.apache.druid.segment.column.ColumnType;
+import org.apache.druid.segment.column.RowSignature;
+import org.apache.druid.segment.column.RowSignature.Builder;
+import org.apache.druid.server.security.Action;
+import org.apache.druid.server.security.Resource;
+import org.apache.druid.server.security.ResourceAction;
+import org.apache.druid.server.security.ResourceType;
+import org.apache.druid.sql.calcite.expression.AuthorizableOperator;
+import org.apache.druid.sql.calcite.expression.DruidExpression;
+import org.apache.druid.sql.calcite.expression.SqlOperatorConversion;
+import org.apache.druid.sql.calcite.planner.DruidSqlValidator;
+import org.apache.druid.sql.calcite.planner.PlannerContext;
+import org.apache.druid.sql.calcite.table.DatasourceMetadata;
+import org.apache.druid.sql.calcite.table.DatasourceTable;
+import org.apache.druid.sql.calcite.table.DatasourceTable.EffectiveMetadata;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+/**
+ * Dynamic table append.
+ *
+ * Enables to use: TABLE(APPEND('t1','t2')); which will provide a union view of the operand tables.
+ */
+public class TableAppendMacro extends SqlUserDefinedTableMacro implements AuthorizableOperator
+{
+
+  public static final OperatorConversion OPERATOR_CONVERSION = new OperatorConversion();
+  public static final SqlOperator APPEND_TABLE_MACRO = new TableAppendMacro();
+
+  private static class OperatorConversion implements SqlOperatorConversion
+  {
+    public static final String FUNCTION_NAME = "APPEND";
+
+    public OperatorConversion()
+    {
+    }
+
+    @Override
+    public SqlOperator calciteOperator()
+    {
+      return APPEND_TABLE_MACRO;
+    }
+
+    @Override
+    public DruidExpression toDruidExpression(PlannerContext plannerContext, RowSignature rowSignature, RexNode rexNode)
+    {
+      throw new IllegalStateException();
+    }
+  }
+
+  private TableAppendMacro()
+  {
+    super(
+        new SqlIdentifier(OperatorConversion.FUNCTION_NAME, SqlParserPos.ZERO),
+        SqlKind.OTHER_FUNCTION,
+        ReturnTypes.CURSOR,
+        null,
+        new OperandMetadata(),
+        null
+    );
+  }
+
+  private static class OperandMetadata implements SqlOperandMetadata
```

Comment:
the `SqlOperandMetadata` seems to say that it is only to be used for fixed number of parameters. do you think is it ok to use to use for variable params as well in this case?

Reply:
discussed offline, the approach taken seems like the only decent way to use `SqlUserDefinedTableMacro` - so we can keep using it. In future, we may want to add support in calcite itself for variadic params in `SqlOperandMetadata` as well